talk/objectorientation/advancedoo.tex

```diff
@@ -84,6 +123,53 @@
   \end{multicols}
 \end{frame}
 
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Polymorphism and object slicing}
+  \begin{block}{What happens if we assign without reference or pointers?}
+    \begin{itemize}
+    \item The data members of the derived object are ``sliced off''
+    \item This is usually wrong
+    \item Pass / use / catch polymorphic objects as references!
+    \end{itemize}
+  \end{block}
+  \begin{multicols}{2}
+    \begin{cppcode*}{gobble=2}
+      Polygon p;
+
+      int f(Drawable d) {...}
+      f(p);  //ok-ish
+
+      try {
+        throw p;
+      } catch (Shape e) {
+        // will be caught
+      }
+    \end{cppcode*}
+    \columnbreak
+    \center
+    \begin{tikzpicture}
+      \memorystack[size x=3cm,word size=1,block size=4,nb blocks=8]
+      \memorypush{...}
+      \memorypush{Shape.a}
+      \memorypush{...}
+      \memorypush{Polygon.nLines}
+      \memorystruct{0}{4}{\tiny Polygon p}
+      \memorypush{Drawable.a}
+      \memorypush{...}
+      \onslide<1>{
+        \memorystruct{5}{6}{\tiny Drawable in f()}
+      }
+      \onslide<2>{
+        \memorypush{Shape.a}
+        \memorypush{...}
+        \memorystruct{5}{8}{\tiny Shape in catch block}
+      }
+    \end{tikzpicture}
+  \end{multicols}
+\end{frame}
```

Comment:
@bernhardmgruber ?
You asked for discussing slicing. Did it go in the direction that you thought?

Reply:
IIRC, a student asked about this during the course, or you explained it. I can't remember exactly.

I think we can remove the `try` and `catch` part of the example. Also:
```c++
      int f(Drawable d) {...}
      f(p);  //ok-ish
```
is not "ok-ish", because if Drawable is abstract, the code would not compile: You cannot create an instance of a Drawable.
Also, if Drawable has any virtual functions, they would no longer be called on the derived class, but on the base class. So you basically lose polymorphism and you are left with the functionality of a plain Drawable.

Maybe the picture of the memory layout is not needed. The main point is that students understand to not do it, and not how it works :)

Reply:
OK, but looking at our discussion, this is clearly too advanced. I wasn't sure that we needed it, just added it because Bernhard wrote it down. Bernhard seems to not be sure why we wrote it down, so maybe the easiest is to really drop the slide.
If the students ask again, we can always pull up godbolt and do it on the fly.

OK to drop it?

Reply:
> Bernhard wrote it down. Bernhard seems to not be sure why we wrote it down

I wrote it down because the discussion came up during the course. One way or another. So I thought: let's add a slide for that!

> maybe the easiest is to really drop the slide.

Maybe. I would prefer to mention slicing somewhere, but I have no time to work on a slide :S So, up to you!

Reply:
OK, removed now! 🙂